### PR TITLE
sql: casting to timestamp should strip tz info

### PR DESCRIPTION
**Pull request by jordanlewis**
And binary operations between timestamp and timestamptz should strip the
tz from the timestamptz.

Confirmed that all of this behavior matches postgres (and that it didn't
before).

Release note (bug fix): correct casts and binary operators between
timestamptz and timestamp in some cases.

### CODE REVIEW

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -820,17 +820,21 @@ var BinOps = map[BinaryOperator]binOpOverload{
 			LeftType:   types.Timestamp,
 			RightType:  types.TimestampTZ,
 			ReturnType: types.Interval,
-			fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
-				nanos := left.(*DTimestamp).Sub(right.(*DTimestampTZ).Time).Nanoseconds()
+			fn: func(ctx *EvalContext, left Datum, right Datum) (Datum, error) {
+				// These two quantities aren't directly comparable. Convert the
+				// TimestampTZ to a timestamp first.
+				nanos := left.(*DTimestamp).Sub(right.(*DTimestampTZ).stripTimeZone(ctx).Time).Nanoseconds()
```
Comment (knz) [inline pkg/sql/sem/tree/eval.go]: please add tests in `eval_test.go` for these comparison function changes.

**Comment by jordanlewis** [comment]
Done, PTAL

**Review by knz — APPROVED** [review]
:lgtm: but please remove the debug printf

> Reviewed 3 of 4 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28128#-:-LIq33V999_p9jabhqCz:b-2whg47)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/sem/tree/datum.go, line 2091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28128#-LIq1hPY0Kr3Z9gczuAc:-LIq1hPY0Kr3Z9gczuAd:b-lfqcw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/946466946c7abecde5887cbb49876b1b13f041eb/pkg/sql/sem/tree/datum.go#L2091)):*
> ```Go
> 	_, locOffset := d.Time.In(ctx.GetLocation()).Zone()
> 	newTime := duration.Add(d.Time.UTC(), duration.FromInt64(int64(locOffset)))
> 	fmt.Println("Stripping", d, locOffset, newTime)
> ```

please remove the `fmt.Println` here

**Comment by jordanlewis** [comment]
Oops, thanks for catching

**Pull request merged**